pkg/runtime/wfengine/activity.go

```
@@ -0,0 +1,298 @@
+	//       introduce some kind of heartbeat protocol to help identify such cases.
+	callback := make(chan bool)
+	wi.Properties[CallbackChannelProperty] = callback
+	if err = a.scheduler.ScheduleActivity(ctx, wi); err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			return newRecoverableError(errors.New(
+				"timed-out trying to schedule an activity execution - this can happen if too many activities are running in parallel or if the workflow engine isn't running"))
+		}
+		return newRecoverableError(fmt.Errorf("failed to schedule an activity execution: %w", err))
+	}
+
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(10 * time.Minute):
+			if deadline, ok := ctx.Deadline(); ok {
+				wfLogger.Warnf("%s: '%s' is still running - will keep waiting until %v", actorID, name, deadline)
+			} else {
+				wfLogger.Warnf("%s: '%s' is still running - will keep waiting indefinitely", actorID, name)
+			}
+		case completed := <-callback:
+			if completed {
+				break loop
+			} else {
+				return newRecoverableError(errExecutionAborted)
+			}
+		}
+	}
+
+	// publish the result back to the workflow actor as a new event to be processed
+	resultData, err := backend.MarshalHistoryEvent(wi.Result)
+	if err != nil {
+		return err
+	}
+	req := invokev1.
+		NewInvokeMethodRequest(AddWorkflowEventMethod).
+		WithActor(WorkflowActorType, workflowID).
+		WithRawDataBytes(resultData).
+		WithContentType(invokev1.OctetStreamContentType)
+	defer req.Close()
+
+	resp, err := a.actorRuntime.Call(ctx, req)
+	if err != nil {
+		return newRecoverableError(fmt.Errorf("failed to invoke '%s' method on workflow actor: %w", AddWorkflowEventMethod, err))
+	}
+	defer resp.Close()
+	return nil
+}
+
+// InvokeTimer implements actors.InternalActor
+func (*activityActor) InvokeTimer(ctx context.Context, actorID string, timerName string, params []byte) error {
+	return nil // timers are not used
```

should we return an error instead? is it ok to give the "impression" that it actually works?

---

Yeah, I can change this to an error.